elrond-wasm/src/api/send_api.rs

```diff
@@ -115,90 +138,104 @@ pub trait SendApi: Clone + Sized {
+            arg_buffer.push_arg(amount);
+            if !data_buf.is_empty() {
+                arg_buffer.push_arg_raw(data_buf);
             }
 
             self.async_call_raw(
                 to,
                 &BigUint::zero(self.type_manager()),
-                serializer.as_slice(),
+                &ManagedBuffer::new_from_bytes(self.type_manager(), ESDT_TRANSFER_STRING),
+                &arg_buffer,
             )
         } else {
-            let mut serializer = HexCallDataSerializer::new(ESDT_NFT_TRANSFER_STRING);
-            serializer.push_argument_bytes(token.to_esdt_identifier().as_slice());
-            serializer.push_argument_bytes(&nonce.to_be_bytes()[..]);
-            serializer.push_argument_bytes(amount.to_bytes_be().as_slice());
-            serializer.push_argument_bytes(to.as_bytes());
-            if !data.is_empty() {
-                serializer.push_argument_bytes(data);
+            let mut arg_buffer = ManagedArgBuffer::new_empty(self.type_manager());
+            arg_buffer.push_arg(token);
+            arg_buffer.push_arg(nonce);
+            arg_buffer.push_arg(amount);
+            arg_buffer.push_arg(to);
+            if !data_buf.is_empty() {
+                arg_buffer.push_arg_raw(data_buf);
             }
 
             self.async_call_raw(
                 &self.blockchain().get_sc_address(),
                 &BigUint::zero(self.type_manager()),
-                serializer.as_slice(),
-            );
+                &ManagedBuffer::new_from_bytes(self.type_manager(), ESDT_NFT_TRANSFER_STRING),
+                &arg_buffer,
+            )
         }
     }
 
-    fn transfer_multiple_esdt_via_async_call(
+    fn transfer_multiple_esdt_via_async_call<D>(
         &self,
-        to: &Address,
-        tokens: &[EsdtTokenPayment<Self::ProxyTypeManager>],
-        data: &[u8],
-    ) -> ! {
-        let mut serializer = HexCallDataSerializer::new(ESDT_MULTI_TRANSFER_STRING);
-        serializer.push_argument_bytes(to.as_bytes());
-        serializer.push_argument_bytes(&tokens.len().to_be_bytes()[..]);
-
-        for token in tokens {
-            serializer.push_argument_bytes(token.token_name.to_esdt_identifier().as_slice());
-            serializer.push_argument_bytes(&token.token_nonce.to_be_bytes()[..]);
-            serializer.push_argument_bytes(token.amount.to_bytes_be().as_slice());
+        to: &ManagedAddress<Self::ProxyTypeManager>,
+        payments: &ManagedVec<Self::ProxyTypeManager, EsdtTokenPayment<Self::ProxyTypeManager>>,
+        data: D,
+    ) -> !
+    where
+        D: ManagedInto<Self::ProxyTypeManager, ManagedBuffer<Self::ProxyTypeManager>>,
+    {
+        let mut arg_buffer = ManagedArgBuffer::new_empty(self.type_manager());
+        arg_buffer.push_arg(to);
+        arg_buffer.push_arg(payments.len());
+
+        for payment in payments.into_iter() {
+            // TODO: check that `!token_name.is_egld()` or let Arwen throw the error?
```

Comment:
I think we should just let it crash "naturally". Should be fairly obvious for the developer that "ESDT MultiTransfer" only works with ESDTs.